SOLR-16462: Create v2 equivalent of v1 'CREATESNAPSHOT', 'LISTSNAPSHOT' and 'DELETESNAPSHOT' (core level)

Jdurham2843 opened the pull request:
https://issues.apache.org/jira/browse/SOLR-16462

<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Solr:

* https://issues.apache.org/jira/projects/SOLR

For something minor (i.e. that wouldn't be worth putting in release notes), you can skip JIRA. 
To create a Jira issue, you will need to create an account there first.

The title of the PR should reference the Jira issue number in the form:

* SOLR-####: <short description of problem or changes>

SOLR must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

This pull request implements V2 equivalent endpoints for the Snapshot API (Create, List, and Delete).

# Solution

I created a SnapshotAPI class that houses new v2 implementations of the CreateSnapshot, ListSnapshot, and DeleteSnapshot operations. The SnapshotAPI endpoints were implemented using the Jersey JAX-RS reference implementation. The original CoreAdminOperation implementations were modified to call the new SnapshotAPI methods. The backup/restore reference documentation has been updated to show both V1 and V2 Snapshot API examples.

# Tests

No new tests were added, but I would be willing to add some if needed. There are existing tests that check that both the coreName and commitName were passed into a given Snapshot operation. Since those values are now part of the URL, I didn't see any reason to test for that against the new V2 methods. 

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

gerlowskija commented on solr/core/src/java/org/apache/solr/handler/admin/api/SnapshotAPI.java:
```diff
@@ -0,0 +1,218 @@
+  }
+
+  public static class CreateSnapshotResponse extends SolrJerseyResponse {
+    @JsonProperty(CoreAdminParams.CORE)
```
[0] It's not required, especially in cases where the nuances of what a particular field represents are unclear, but if you DO have a good idea for any of these fields and are willing, you can use the `@Schema` annotation to document it.

These descriptions end up in the OpenAPI spec generated by the Solr build, which we'll use in subsequent releases to create Solr clients for different languages (e.g. golang, Python).  We're also hoping to use them to auto-generate certain sections of our ref-guide.  So lots of indirect value there if you're able to document any of these.

Jdurham2843 replied:
I'll make an effort to add some Schema statements. It seems like a good idea to add those now while we're working on this API.

gerlowskija commented on solr/core/src/java/org/apache/solr/handler/admin/DeleteSnapshotOp.java:
```diff
@@ -17,35 +17,26 @@
+    final SnapshotAPI.DeleteSnapshotResponse response =
+        snapshotAPI.deleteSnapshot(coreName, commitName);
+
+    it.rsp.add(CoreAdminParams.CORE, response.coreName);
```
[Q] ditto, re: can we use `V2ApiUtils.squashIntoSolrResponseWithoutHeader` here instead?

gerlowskija commented on solr/core/src/java/org/apache/solr/handler/admin/api/SnapshotAPI.java:
```diff
@@ -0,0 +1,218 @@
+
+    try {
+      core.deleteNamedSnapshot(snapshotName);
+      // Ideally we shouldn't need this. This is added since the RPC logic in
```
[0] I know this comment was there in the original code that you moved from DeleteSnapshotOp, so this isn't a comment on your PR, but it makes no sense to me.

What is the "this" being referred to?  Adding the coreName to the response?

Anyway, just talking out loud I guess....

gerlowskija commented on solr/solr-ref-guide/modules/deployment-guide/pages/backup-restore.adoc:
```diff
@@ -234,38 +234,43 @@ http://localhost:8983/solr/gettingstarted/replication?command=restorestatus&wt=x
-|===
-+
-The name to store the snapshot as.
+With the v2 API, the core and snapshot names are part of the path, no longer query parameters.
```
[+1] Awesome docs, thanks for adding these!  Big fan of the use of "tabs" in the docs to handle v1/v2 examples.

Only nitpick I'd make here is to drop or rephrase the "no longer query parameters" bit (maybe use "instead of"?).  But even that's probably personal preference.

Jdurham2843 replied:
No problem! I actually just copied the structure from elsewhere in the docs, so it wasn't much work on my part. I did go ahead and change the wording since I thought the "instead of" read better.

gerlowskija reviewed (changes requested):
Everything here looks great; thanks John!

I think we will want some tests here, and I know we still have the "async" question to discuss.  But I'm not sure yet what either of those should look like, so I figured I'd leave a few "initial impression" comments while I mull those bigger questions over a bit.

Jdurham2843 commented on solr/core/src/java/org/apache/solr/handler/admin/CoreAdminOperation.java:
```diff
@@ -278,31 +276,28 @@ public enum CoreAdminOperation implements CoreAdminOp {
-            }
-          }
-          it.rsp.add(SolrSnapshotManager.SNAPSHOTS_INFO, result);
+        final NamedList<Object> snapshotsResult = new NamedList<>();
```
I hadn't seen that method before, but that seems like the better approach! I refactored all three operations to use it with success on all but the v1 ListSnapshots operation. From what I've seen, the original method returns a response with the following structure:
```
{
	"responseHeader": {
		"status": 0,
		"QTime": 0
	},
	"snapshots": [
		"snapshot4",
		[
			"generation",
			"2",
			"indexDirPath",
			"C:\\<redacted-path>\\solr\\solr\\packaging\\build\\dev\\example\\techproducts\\solr\\techproducts\\data\\index/"
		],
		"snapshot3",
		[
			"generation",
			"2",
			"indexDirPath",
			"C:\\<redacted-path>\\solr\\solr\\packaging\\build\\dev\\example\\techproducts\\solr\\techproducts\\data\\index/"
		],
		"snapshot2",
		[
			"generation",
			"2",
			"indexDirPath",
			"C:\\<redacted-path>\\solr\\solr\\packaging\\build\\dev\\example\\techproducts\\solr\\techproducts\\data\\index/"
		],
		"snapshot6",
		[
			"generation",
			"2",
			"indexDirPath",
			"C:\\<redacted-path>\\solr\\solr\\packaging\\build\\dev\\example\\techproducts\\solr\\techproducts\\data\\index/"
		]
	]
}
```

While using the squash method returns the following structure:
```
{
	"responseHeader": {
		"status": 0,
		"QTime": 7
	},
	"snapshots": {
		"snapshot4": {
			"generation": 2,
			"indexDirPath": "C:\\<redacted-path>\\solr\\solr\\packaging\\build\\dev\\example\\techproducts\\solr\\techproducts\\data\\index/"
		},
		"snapshot3": {
			"generation": 2,
			"indexDirPath": "C:\\<redacted-path>\\solr\\solr\\packaging\\build\\dev\\example\\techproducts\\solr\\techproducts\\data\\index/"
		},
		"snapshot2": {
			"generation": 2,
			"indexDirPath": "C:\\<redacted-path>\\solr\\solr\\packaging\\build\\dev\\example\\techproducts\\solr\\techproducts\\data\\index/"
		},
		"snapshot6": {
			"generation": 2,
			"indexDirPath": "C:\\<redacted-path>\\solr\\solr\\packaging\\build\\dev\\example\\techproducts\\solr\\techproducts\\data\\index/"
		}
	}
}
```

Would the change in the structure of the response for the V1 endpoint be considered a breaking change that we should prevent?

gerlowskija replied:
Oof.  I really don't want to defend that original v1 format, which looks clearly wrong to me.  But I think you're right to suspect that changing this would be a "breaking change" that we should prevent 😞

gerlowskija replied:
After giving this a bit more thought, it'd be a real shame to cement this behavior, which is probably a bug in any case.  If it's an unintentional bug, then it doesn't fall under Solr's standard backcompat policy: we're allowed to fix bugs after all.

I started a discussion about this on the 'dev@' mailing list.  Hopefully someone else will chime in there with agreement, at which point I'm happy to go forward with the corrected syntax your PR uses.

gerlowskija replied:
There were no objections on the [dev@ mailing list thread](https://lists.apache.org/thread/37vcxnprxlrnnqf8wnoms4f66xv0mb85), so I think we're fine to treat this as a bug and not a breaking change.

Jdurham2843 commented on solr/core/src/java/org/apache/solr/handler/admin/api/SnapshotAPI.java:
```diff
@@ -0,0 +1,218 @@
+import org.apache.solr.jersey.SolrJerseyResponse;
+
+@Path("/cores/{coreName}/snapshots")
+public class SnapshotAPI extends JerseyResource {
```
That's a good idea - I'll change the name!

Jdurham2843 commented:
I believe I've covered most of your first round of comments @gerlowskija , aside from tests and the uses of the squash method on the ListSnapshotAPI. I also responded to your thoughts on the async handling code back on the jira ticket as well. I'll work to add some tests focused on this new CoreSnapshotAPI class while we work through the our async game plan.

gerlowskija commented:
Awesome; thanks John.  I replied on JIRA, about the async stuff as well, so I think that's everything.  Lmk if I missed a question though!

Feel free to push up whatever you've got from time to time if you're interested in a sanity check on the  "delta", even if it's not quite "finished" yet.

gerlowskija commented:
Recent changes (`@Schema` documentation, failure-case unit tests, API class renaming) all look great.

To summarize where we are, I think there's two remaining items to clear up before this can be merged:

1. A decision on whether we should hold strict backcompat for the v1 LISTSNAPSHOT response format, or go with the (clearly preferable) syntax the v2 API has currently.  I opened a dev@ thread about that, which slows this PR down a bit, but it's better to get consensus on this rather than risk someone complaining later.  If there's no objections by mid-next week I say we go forward with the improved syntax.
2. Handling of the async param.  I think I offered to help with this in JIRA; so I'll take a crack at it if you haven't already started in on it @Jdurham2843 ?

Jdurham2843 commented:
1. Awesome, sounds good! The new syntax definitely looks better (and may be easier to parse if someone were using the API from another program), but it's probably a good idea to get consensus first.
2. If we're still waiting on item 1, I'll go ahead and take a stab at it. Also, in reference to the your jira response, I'd be interested in working together on a more general solution too. It sounds like a good way to get familiar with other APIs in Solr.
3. I added three tests just checking the errors response from CoreSnapshotAPI. I started to write some tests (heavily mocked) to verify the response output, but ran into an issue where I couldn't mock the IndexDeletionPolicyWrapper since its a final class. Are there any additional things you can think of that would be worth testing? My current thinking is that we're just trying to test/validate the interface (request, response, errors).

Jdurham2843 commented:
1. (pending)
2. I created a new CoreAdminAPIBase class to house the async handling code. The class takes the async code from CoreAdminHandler and wraps it in handle method. The handle method then takes it some parameters from each endpoint along with a Supplier<T extends JerseyResponse>. This appears to work the same as the original V1 APIs async code. However, injecting a SolrQueryRequest and SolrQueryResponse has led to a MultiException stack error. See below for the output.
3. I figured out a way to write the response validation tests without needing to mock. It turns out everything I needed was already in the Test's base class. I just had to call a few methods to setup the TestHarness.

Let me know what you think!

MultiException error when calling new V2 methods:
```
Nov 06, 2022 7:01:24 PM org.glassfish.jersey.internal.Errors logErrors
WARNING: The following warnings have been detected: WARNING: HK2 failure has been detected in a code that does not run in an active Jers
ey Error scope.
WARNING: Unknown HK2 failure detected:
MultiException stack 1 of 3
java.lang.NullPointerException
        at org.glassfish.jersey.server.internal.process.RequestProcessingConfigurator$ContainerRequestFactory.get(RequestProcessingConfi
gurator.java:67)
        at org.glassfish.jersey.server.internal.process.RequestProcessingConfigurator$ContainerRequestFactory.get(RequestProcessingConfi
gurator.java:56)
        at org.glassfish.jersey.inject.hk2.SupplierFactoryBridge.provide(SupplierFactoryBridge.java:76)
        at org.jvnet.hk2.internal.FactoryCreator.create(FactoryCreator.java:129)
        at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:463)
        at org.glassfish.jersey.inject.hk2.RequestContext.findOrCreate(RequestContext.java:59)
        at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:93)
        at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:679)
        at org.jvnet.hk2.internal.ThreeThirtyResolver.resolve(ThreeThirtyResolver.java:54)
        at org.jvnet.hk2.internal.ClazzCreator.resolve(ClazzCreator.java:188)
        at org.jvnet.hk2.internal.ClazzCreator.resolveAllDependencies(ClazzCreator.java:205)
        at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:334)
        at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:463)
        at org.jvnet.hk2.internal.PerLookupContext.findOrCreate(PerLookupContext.java:46)
        at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:93)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:67)
        at org.jvnet.hk2.internal.FactoryCreator.dispose(FactoryCreator.java:151)
        at org.jvnet.hk2.internal.SystemDescriptor.dispose(SystemDescriptor.java:518)
        at org.glassfish.jersey.inject.hk2.RequestContext.lambda$findOrCreate$0(RequestContext.java:60)
        at org.glassfish.jersey.internal.inject.ForeignDescriptorImpl.dispose(ForeignDescriptorImpl.java:63)
        at org.glassfish.jersey.inject.hk2.Hk2RequestScope$Instance.remove(Hk2RequestScope.java:126)
        at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        at org.glassfish.jersey.inject.hk2.Hk2RequestScope$Instance.release(Hk2RequestScope.java:143)
        at org.glassfish.jersey.process.internal.RequestScope.release(RequestScope.java:246)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:267)
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684)
        at org.apache.solr.api.V2HttpCall.invokeJerseyRequest(V2HttpCall.java:371)
        at org.apache.solr.api.V2HttpCall.invokeJerseyRequest(V2HttpCall.java:338)
        at org.apache.solr.api.V2HttpCall.executeCoreRequest(V2HttpCall.java:452)
        at org.apache.solr.servlet.HttpSolrCall.call(HttpSolrCall.java:554)
        at org.apache.solr.servlet.SolrDispatchFilter.dispatch(SolrDispatchFilter.java:251)
        at org.apache.solr.servlet.SolrDispatchFilter.lambda$doFilter$0(SolrDispatchFilter.java:219)
        at org.apache.solr.servlet.ServletUtils.traceHttpRequestExecution2(ServletUtils.java:257)
        at org.apache.solr.servlet.ServletUtils.rateLimitRequest(ServletUtils.java:227)
        at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:214)
        at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:196)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:201)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:552)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
        at org.eclipse.jetty.server.handler.InetAccessHandler.handle(InetAccessHandler.java:177)
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.rewrite.handler.RewriteHandler.handle(RewriteHandler.java:322)
        at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:772)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.Server.handle(Server.java:516)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
        at java.base/java.lang.Thread.run(Thread.java:829)
MultiException stack 2 of 3
java.lang.IllegalArgumentException: While attempting to resolve the dependencies of org.apache.solr.jersey.InjectionFactories$SolrQueryR
equestFactory errors were found
        at org.jvnet.hk2.internal.ClazzCreator.resolveAllDependencies(ClazzCreator.java:224)
        at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:334)
        at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:463)
        at org.jvnet.hk2.internal.PerLookupContext.findOrCreate(PerLookupContext.java:46)
        at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:93)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:67)
        at org.jvnet.hk2.internal.FactoryCreator.dispose(FactoryCreator.java:151)
        at org.jvnet.hk2.internal.SystemDescriptor.dispose(SystemDescriptor.java:518)
        at org.glassfish.jersey.inject.hk2.RequestContext.lambda$findOrCreate$0(RequestContext.java:60)
        at org.glassfish.jersey.internal.inject.ForeignDescriptorImpl.dispose(ForeignDescriptorImpl.java:63)
        at org.glassfish.jersey.inject.hk2.Hk2RequestScope$Instance.remove(Hk2RequestScope.java:126)
        at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        at org.glassfish.jersey.inject.hk2.Hk2RequestScope$Instance.release(Hk2RequestScope.java:143)
        at org.glassfish.jersey.process.internal.RequestScope.release(RequestScope.java:246)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:267)
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684)
        at org.apache.solr.api.V2HttpCall.invokeJerseyRequest(V2HttpCall.java:371)
        at org.apache.solr.api.V2HttpCall.invokeJerseyRequest(V2HttpCall.java:338)
        at org.apache.solr.api.V2HttpCall.executeCoreRequest(V2HttpCall.java:452)
        at org.apache.solr.servlet.HttpSolrCall.call(HttpSolrCall.java:554)
        at org.apache.solr.servlet.SolrDispatchFilter.dispatch(SolrDispatchFilter.java:251)
        at org.apache.solr.servlet.SolrDispatchFilter.lambda$doFilter$0(SolrDispatchFilter.java:219)
        at org.apache.solr.servlet.ServletUtils.traceHttpRequestExecution2(ServletUtils.java:257)
        at org.apache.solr.servlet.ServletUtils.rateLimitRequest(ServletUtils.java:227)
        at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:214)
        at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:196)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:201)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:552)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
        at org.eclipse.jetty.server.handler.InetAccessHandler.handle(InetAccessHandler.java:177)
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.rewrite.handler.RewriteHandler.handle(RewriteHandler.java:322)
        at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:772)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.Server.handle(Server.java:516)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
        at java.base/java.lang.Thread.run(Thread.java:829)
MultiException stack 3 of 3
java.lang.IllegalStateException: Unable to perform operation: resolve on org.apache.solr.jersey.InjectionFactories$SolrQueryRequestFacto
ry
        at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:363)
        at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:463)
        at org.jvnet.hk2.internal.PerLookupContext.findOrCreate(PerLookupContext.java:46)
        at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:93)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:67)
        at org.jvnet.hk2.internal.FactoryCreator.dispose(FactoryCreator.java:151)
        at org.jvnet.hk2.internal.SystemDescriptor.dispose(SystemDescriptor.java:518)
        at org.glassfish.jersey.inject.hk2.RequestContext.lambda$findOrCreate$0(RequestContext.java:60)
        at org.glassfish.jersey.internal.inject.ForeignDescriptorImpl.dispose(ForeignDescriptorImpl.java:63)
        at org.glassfish.jersey.inject.hk2.Hk2RequestScope$Instance.remove(Hk2RequestScope.java:126)
        at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        at org.glassfish.jersey.inject.hk2.Hk2RequestScope$Instance.release(Hk2RequestScope.java:143)
        at org.glassfish.jersey.process.internal.RequestScope.release(RequestScope.java:246)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:267)
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684)
        at org.apache.solr.api.V2HttpCall.invokeJerseyRequest(V2HttpCall.java:371)
        at org.apache.solr.api.V2HttpCall.invokeJerseyRequest(V2HttpCall.java:338)
        at org.apache.solr.api.V2HttpCall.executeCoreRequest(V2HttpCall.java:452)
        at org.apache.solr.servlet.HttpSolrCall.call(HttpSolrCall.java:554)
        at org.apache.solr.servlet.SolrDispatchFilter.dispatch(SolrDispatchFilter.java:251)
        at org.apache.solr.servlet.SolrDispatchFilter.lambda$doFilter$0(SolrDispatchFilter.java:219)
        at org.apache.solr.servlet.ServletUtils.traceHttpRequestExecution2(ServletUtils.java:257)
        at org.apache.solr.servlet.ServletUtils.rateLimitRequest(ServletUtils.java:227)
        at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:214)
        at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:196)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:201)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:552)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
        at org.eclipse.jetty.server.handler.InetAccessHandler.handle(InetAccessHandler.java:177)
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.rewrite.handler.RewriteHandler.handle(RewriteHandler.java:322)
        at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:772)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.Server.handle(Server.java:516)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

gerlowskija commented:
> If we're still waiting on item 1, I'll go ahead and take a stab at the async code.

Alright, thanks for grabbing that!  On item (1), I think the conclusion on that `dev@` thread was that we're good to change the appearance here, since this very much looks like a bug.

I'll take a look sometime today to review what you've added.

gerlowskija commented on solr/core/src/test/org/apache/solr/handler/admin/api/CoreSnapshotAPITest.java:
```diff
@@ -0,0 +1,161 @@
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class CoreSnapshotAPITest extends SolrTestCaseJ4 {
```
[+1] These tests look awesome!  Nice to avoid heavy mocking and also avoid standing up a full SolrCloud cluster or anything similarly-heavy like that.

gerlowskija commented on solr/core/src/java/org/apache/solr/handler/admin/api/CoreAdminAPIBase.java:
```diff
@@ -0,0 +1,210 @@
+
+public abstract class CoreAdminAPIBase extends JerseyResource {
+
+  private static final ExecutorService PARALLEL_EXECUTOR =
```
[-1] I love the approach on this class overall (esp. the `handle` method that takes in a Supplier), but I'm concerned about reliance on static state here.

I hinted at this a bit on JIRA, but should've been more explicit:

> I guess we could create a base-class for our APIs, e.g. "AsyncTrackingApiBase", with a static instance to hold this status info, that gets extended by various API subclasses. I don't love the idea of introducing non-constant static state, but I think it'd work?

It's my fault for not being more explicit: it's not the norm, but it's also not unheard of to run multiple Solr instances from within the same JVM.  Solr's tests do it, for one, and I've heard of actual applications doing this too on occasion via EmbeddedSolrServer and similar classes.

Maybe this would "just work" if the tests and EmbeddedSolrServer know to use different classloaders, but if it doesn't, then the use of static state could cause tricky bugs - e.g. the same executor being used for multiple Solr instances.

I'll run the test suite and poke around some of the related code to double-check my fears here, but assuming they're well founded, we'll need a different approach here.

One approach that's only slightly a "dodge" would be to essentially re-use the same status-map and `parallelExecutor` that CoreAdminHandler creates.  (i.e. Add getters for these things to `CoreAdminHandler` and then modify the Jersey application to make the status-map and executor injectable as singletons, similar to what we do for CoreContainer [here](https://github.com/apache/solr/blob/c86128855807042e25c5bd320e6f7729784e8884/solr/core/src/java/org/apache/solr/core/CoreContainer.java#L1067))

This would have a few side-benefits:

- avoids the creation of an additional executor (the current PR creates one in the handler and one in this API class)
- no executor lifecycle changes required, since we're reusing the executor
- injection gives the API classes enough indirection that they won't require any changes when we eventually nuke v1 and need to find a different way of managing the executor/status-map
- no duplication of constants and status-map init logic (current PR has this code in both the handler and this API class).

Any thoughts on that?  I'm happy to make the changes required as the rework is largely my fault for not communicating my concerns around static-state strongly enough.

Jdurham2843 replied:
Oh gotcha, not a problem - The way I read your original comment, I thought we were going fully with the static approach first, I can take some blame for my interpretation. I think what you're laying out here makes sense and I hadn't considered the case of multiple Solr Contexts running on a single VM. I took a little bit of time to see how I would register the map/executor, but haven't made any code changes yet. I will attempt to work on it this weekend if you haven't already gotten to it.

Would migrating the status map and executor into a class that can then be used by both the CoreAdminHandler and made injectable for the CoreAdminAPIBase work? That may save us from having to inject multiple things and could potentially get us closer to the future state of what async could look like.

gerlowskija commented on solr/core/src/java/org/apache/solr/handler/admin/api/CoreAdminAPIBase.java:
```diff
@@ -0,0 +1,210 @@
+import org.apache.solr.util.tracing.TraceUtils;
+import org.slf4j.MDC;
+
+public abstract class CoreAdminAPIBase extends JerseyResource {
```
[0] Small nit, should add Javadocs explaining what this base class handles for callers and when/how it should be used.

Jdurham2843 replied:
That sounds like a good idea, I'll see about adding those!

gerlowskija reviewed (changes requested):
I love the base-class approach overall, but I'm leery of the static state that CoreAdminAPIBase introduces.  I proposed a fix for that (see my comment in line); let me know what you think when you get a chance.  I'll aim to push up commit in a few days unless you (1) don't like the alternative I suggested, or (2) have some interest in digging into the JAX-RS/Jersey code personally.

Jdurham2843 commented on solr/core/src/java/org/apache/solr/handler/admin/CoreAdminOperation.java:
```diff
@@ -278,31 +282,16 @@ public enum CoreAdminOperation implements CoreAdminOp {
-            throw new SolrException(ErrorCode.BAD_REQUEST, "Unable to locate core " + cname);
-          }
+        final CoreSnapshotAPI.ListSnapshotsResponse response =
+            coreSnapshotAPI.listSnapshots(coreName, null);
```
fixed

Jdurham2843 commented:
Update:
1. I added some JavaDocs to CoreAdminAPIBase and CoreSnapshotAPI. I tried to follow the same general patterns I saw in JerseyResource and in the SchemaNameAPI class.
2. I made an attempt at reusing the executor and statusMap from CoreAdminHandler in the CoreAdminAPIBase class. Instead of Injecting both separately, I moved them both into a static nested class on CoreAdminHandler and wrote some code to Inject that. Even if we decide not to use this version and instead you have a better approach, we can revert these changes. I at least wanted to give it a shot first!
3. I added the V2 squash static method to the V1 ListSnapshots operation.

gerlowskija commented:
Hi @Jdurham2843 

Just wanted to give you a short update on my lack of progress on this.

It is definitely still on my radar, but I've been a little distracted recently focusing on a performance regression that someone noticed in the JAX-RS framework overall: another committer noticed that the recent addition of the JAX-RS framework has slowed down Solr startup, in some cases quite significantly.  That's turned into a big blocker for the JAX-RS work overall: there's even talk of deferring the release of our JAX-RS v2 APIs until that perf regression can be mitigated.

So most of my focus has shifted to that - after all, it's not much good focusing on individual JAX-RS APIs if shipping the whole framework gets delayed 😛 If you're curious, see SOLR-16531 for more details on all this.

Anyway, that's all to say that I haven't gotten to this PR, but that I haven't forgotten it or let it fall off my radar.  I'm just focusing on this performance blocker, and plan to return to this when that's finished.  I appreciate all your work here and definitely haven't forgotten it!

Jdurham2843 commented:
That doesn't sound like any fun, but I totally get it and no worries! This work should be here whenever your in a position to look at it. Hopefully the regression isn't too hard to remediate.

gerlowskija commented:
Alright, my apologies again for the big delay @Jdurham2843, but it looks like SOLR-16531 is near enough to closing out that I'm ready to pick this up again.

Re-reading the conversation here, it sounds like you had addressed all of the sticking points (particularly the executor injection and async handling), and it's ready for a re-review?  Lmk if I've overlooked some detail, but otherwise I'll try to take a look in the next workday or two!

Jdurham2843 commented:
Yep, I think it should be ready for review! I made an attempt to handle async requests, but if you decide there's a better way to do it, I can look at reverting those changes. I also went ahead and did a rebase to bring the branch up to date with main as well.

gerlowskija commented on solr/core/src/java/org/apache/solr/handler/admin/api/CoreAdminAPIBase.java:
```diff
@@ -0,0 +1,157 @@
+   * @param supplier the work that the calling method wants done.
+   * @return the supplied T solrJerseyResponse
+   */
+  public <T extends SolrJerseyResponse> T handle(
```
[+1] Overall I really like how this approach panned out; very cool!

[0] Small nit: can you think of a name here that'd make the method's purpose clearer than `handle`?  I'm struggling a bit, but 'handle' feels generic. Maybe `handlePotentiallyAsynchronousTask`?

[-0] The one downside of this approach is that CoreAdminHandler still has a lot of code that looks similar to what's in this method.  (Specifically lines 200-250ish in CoreAdminHandler.)  Maybe this is as good as it gets, but can you think of any way we could refactor CoreAdminHandler, or this method, or both to cut down on the duplication between the two?

gerlowskija reviewed (commented):
This all looks awesome!  Left a few comments inline.

My biggest concern at this point is that there's some duplicated control-flow in CoreAdminAPIBase and CoreAdminHandler that it feels like we should be able to refactor out, though I haven't sat down long enough to figure out how.

That's small in the grand scheme of things though, so I'm going to start testing locally tomorrow in preparation to commit.

Are you still seeing any of the injection errors you mentioned earlier on, or were you able to find the root cause of those?

Jdurham2843 commented:
Hey! I'll start testing the changes you made, and see if the injection errors are still around. From a first look at the changes, I agree with all of them. Regarding the near-duplicate async processing code: I remember trying to dedupe it, but hit a point where I couldn't see how to do it without creating a mess. I'm glad you found a way to make it work, because it felt kind of wrong to leave that in the codebase.

Jdurham2843 commented:
I did some testing and things appeared to be working correctly. I am still seeing the injection exception. Any thoughts what that might be?

gerlowskija commented:
Hmm, I haven't noticed that Injection error in the logs, though maybe I missed it since the API returns successfully?

I'll play around with it some more and pay more attention to the logs on successful API calls. 🤔

Jdurham2843 commented:
From what I can tell, the injection exception isn't making it into any logs. Instead its being spat out into the terminal where I ran the solr start command. I'm only seeing this error when running in Windows (Powershell or Command Prompt). If I run it under Ubuntu (wsl2), I don't see the error at all. Maybe something OS specific?

gerlowskija commented:
> I'm only seeing this error when running in Windows

Huh, that's a weird wrinkle.  There's nothing OS-specific (that I'm aware of) in any of the dependency-injection stuff.  And, fwiw, I took another stab at reproducing on MacOS this morning and couldn't see any of the logs either.

If I had to guess, I'd say that this is probably related to how Solr uses separate Jersey applications for "container" and "per-core" APIs.  There are some situations where Solr isn't sure what "type" an incoming request is, so it tries sending the request to one Jersey app, and only tries the second if the first errors out.  That'd help explain how we see clear errors in the logs despite the request ultimately succeeding: one Jersey app fails with the observed errors before the second Jersey app succeeds.

But I'm stumped why this behavior would be OS dependent, or even why it'd fail with an injection error.  In the scenario above the first Jersey app typically fails with a NotFoundException that is handled specially so that it doesn't pollute the logs.

----

In terms of moving forward here, if the requests failed on Windows it'd be a very different matter.  But I think the extra noise in the logs is probably tolerable for now given that v2 is an experimental feature.  Absolutely something that needs fixed, but it feels like a framework problem and not something inherent to this PR.

I'll plan on merging this to 'main' and letting it bake a bit before backporting.

Jdurham2843 commented:
Sounds good, thanks for working with me on this!
